File: charrss.py
"""
charrss.py  -- Reverse Seed Search for Melee's random *character* selection.

The canonical runtime algorithm is a direct truncated-LCG / Hidden-Number-Problem
(HNP) reconstruction: given N observed character IDs (ints 0..24), it recovers the
32-bit RNG anchor seed via a single Babai closest-vector solve against an
LLL-reduced lattice. Functions: cvp_basis / cvp_prepare / cvp_search.

Character-RSS specifics:
  - Each character consumes TWO LCG rolls: char = rand_int(s,25), then one unused
    roll. So consecutive characters are two LCG steps apart; the relevant advance
    map is next2 = next ∘ next (multiplier A², not A).
  - bound = 25 (characters), not 145 (tags). Each character pins the anchor to a
    1/25 interval of the 32-bit seed space (see RSS_IMPLEMENTATION.md §1–§2).

Author: btt-manip-backend
"""

import random as _random

from fractions import Fraction as F

MASK = (1 << 32) - 1
SIZE = 1 << 32
MAXINT = MASK
HSIZE = 65536

A = 214013          # LCG multiplier
C = 2531011         # LCG increment
BOUND = 25          # number of selectable characters

# next2 = one *character* step = two LCG steps
A2 = (A * A) & MASK
C2 = (A * C + C) & MASK
# inverse of next2 (for mapping an ending seed back one character step)
A_INV = pow(A, -1, SIZE)
A2_INV = (A_INV * A_INV) & MASK


def next(seed):
    return (seed * A + C) & MASK


def next2(seed):
    return (seed * A2 + C2) & MASK


def prev2(seed):
    return ((seed - C2) * A2_INV) & MASK


def rand_int(seed, bound=BOUND):
    return ((seed >> 16) * bound) >> 16


# ---------------------------------------------------------------------------
# Affine coefficients of next2^k :  next2^k(s) = A2^k * s + B2[k]   (mod 2^32)
# ---------------------------------------------------------------------------
def char_affine(n):
    a = [1]
    b = [0]
    for _ in range(1, n + 1):
        a.append((a[-1] * A2) & MASK)
        b.append((b[-1] * A2 + C2) & MASK)
    return a, b


# ===========================================================================
# Lattice machinery: LLL reduction and coefficient-vector construction for
# the character seed-recovery problem.
# ===========================================================================
def _lll(B, delta=F(3, 4)):
    B = [[F(x) for x in row] for row in B]
    n = len(B)

    def dot(u, v):
        return sum(x * y for x, y in zip(u, v))

    def gram():
        Bs, mu = [], [[F(0)] * n for _ in range(n)]
        for i in range(n):
            bi = B[i][:]
            for j in range(i):
                mu[i][j] = dot(B[i], Bs[j]) / dot(Bs[j], Bs[j])
                bi = [x - mu[i][j] * y for x, y in zip(bi, Bs[j])]
            Bs.append(bi)
        return Bs, mu

    Bs, mu = gram()
    k = 1
    while k < n:
        for j in range(k - 1, -1, -1):
            if abs(mu[k][j]) > F(1, 2):
                q = round(mu[k][j])
                B[k] = [x - q * y for x, y in zip(B[k], B[j])]
                Bs, mu = gram()
        if dot(Bs[k], Bs[k]) >= (delta - mu[k][k - 1] ** 2) * dot(Bs[k - 1], Bs[k - 1]):
            k += 1
        else:
            B[k], B[k - 1] = B[k - 1], B[k]
            Bs, mu = gram()
            k = max(k - 1, 1)
    return [[int(x) for x in row] for row in B]


# ===========================================================================
# Direct CVP reconstruction (truncated-LCG / Hidden-Number-Problem solver)
# ---------------------------------------------------------------------------
# The shipped runtime algorithm (ported to charrss.js). Each character pins
# s_k = a_k*u + b_k (mod 2^32) to a known interval; we recover the anchor u
# with a single Babai nearest-plane closest-vector solve. Validated
# 5000/5000 at 9-10 chars with no brute-force residual (RSS_IMPLEMENTATION.md).
# ===========================================================================
def cvp_basis(n):
    """LLL-reduced basis R (n x n) of the HNP lattice for an n-character search:
    row 0 = (a_0..a_{n-1}) (coeff of the anchor u; a_0=1), rows 1..n-1 = 2^32*e_k.
    Depends only on n, so it is generated offline once and baked."""
    a, _ = char_affine(n)
    avec = a[:n]
    G = [avec[:]]
    for k in range(1, n):
        row = [0] * n
        row[k] = SIZE
        G.append(row)
    return _lll(G)


def cvp_gram_schmidt(R):
    """Exact-rational Gram-Schmidt of the reduced basis (b*_i and <b*_i,b*_i>)."""
    n = len(R)
    Rf = [[F(x) for x in row] for row in R]

    def dot(u, v):
        return sum(x * y for x, y in zip(u, v))

    Bs, den = [], []
    for i in range(n):
        bi = Rf[i][:]
        for j, bs in enumerate(Bs):
            bi = [x - (dot(Rf[i], bs) / den[j]) * y for x, y in zip(bi, bs)]
        Bs.append(bi)
        den.append(dot(bi, bi))
    return Bs, den


def cvp_prepare(R):
    """Precompute the float Gram-Schmidt data the enumeration needs. Returns a
    dict that is also exactly what gets baked into charrss_constants.js:
      basis  -- R (int), gsStar -- b*_i (float), gsDen -- ||b*_i||^2 (float),
      mu     -- mu[j][i] = <R[j],b*_i>/||b*_i||^2 (float), r0 -- column 0 of R.
    Lattice arithmetic is float; correctness comes from the exact regenerate-and-
    verify at each enumeration leaf, so double precision is sufficient."""
    n = len(R)
    Bs_f, den_f = cvp_gram_schmidt(R)
    gs_star = [[float(x) for x in row] for row in Bs_f]
    gs_den = [float(d) for d in den_f]
    mu = [[0.0] * n for _ in range(n)]
    for j in range(n):
        for i in range(j):
            mu[j][i] = sum(R[j][k] * gs_star[i][k] for k in range(n)) / gs_den[i]
    return {"n": n, "basis": R, "gsStar": gs_star, "gsDen": gs_den, "mu": mu,
            "r0": [R[i][0] for i in range(n)]}


def cvp_search(chars, prep, pad=1.5):
    """Recover ALL anchor seeds u (rand_int(u,25)==chars[0]) that reproduce the
    character sequence, via Schnorr-Euchner closest-vector enumeration with a
    radius covering the interval half-widths. Every enumerated candidate is
    verified by regenerating the sequence (so no false positives, and float
    lattice math is safe). Returns a sorted list (normally a single anchor)."""
    n = len(chars)
    gs_star, gs_den, mu, r0 = prep["gsStar"], prep["gsDen"], prep["mu"], prep["r0"]
    _, b = char_affine(n)
    bvec = b[:n]

    t = [0.0] * n
    halfw2 = 0.0
    for k, v in enumerate(chars):
        lo, up = get_l_and_u_bounds(v)
        t[k] = float(((lo + up) // 2 - bvec[k]) & MASK)
        hw = (up - lo) / 2.0
        halfw2 += hw * hw
    r2 = halfw2 * pad
    tstar = [sum(t[k] * gs_star[i][k] for k in range(n)) / gs_den[i] for i in range(n)]

    x = [0] * n
    found = set()

    def rec(i, dacc):
        if i < 0:
            u = 0
            for k in range(n):
                u += x[k] * r0[k]
            u &= MASK
            if generate_chars(u, n) == chars:
                found.add(u)
            return
        s = 0.0
        for j in range(i + 1, n):
            s += x[j] * mu[j][i]
        center = tstar[i] - s
        rem = r2 - dacc
        if rem < 0:
            return
        bound = (rem / gs_den[i]) ** 0.5
        lo_x = int(center - bound) - 1
        hi_x = int(center + bound) + 1
        for xi in range(lo_x, hi_x + 1):
            d = xi - center
            dnew = dacc + gs_den[i] * d * d
            if dnew <= r2:
                x[i] = xi
                rec(i - 1, dnew)

    rec(n - 1, 0.0)
    return sorted(found)


# ===========================================================================
# Interval helpers (bound defaulted to 25)
# ===========================================================================
def lower_bound(val, bound=BOUND):
    return int(HSIZE / bound * val + 1) * HSIZE


def get_l_and_u_bounds(val, bound=BOUND):
    if val == 0:
        return 0, lower_bound(1, bound) - 1
    elif val == bound - 1:
        return lower_bound(val, bound), MAXINT
    else:
        return lower_bound(val, bound), lower_bound(val + 1, bound) - 1


# ---- reference generators and shared validation ----
def validate_cvp(prep, num_chars, trials, rng_seed=12345, extra_seeds=()):
    """Run cvp_search over extra_seeds then trials random seeds and check
    correctness. Returns a dict:
      found, not_found, false_pos, multi  -- counts
      failures                            -- list of human-readable failure strings
    extra_seeds are tested first and tagged 'edge' in failure messages;
    random seeds are tagged 'random'."""
    rng = _random.Random(rng_seed)
    found = not_found = false_pos = multi = 0
    failures = []
    extra_set = set(extra_seeds)
    all_seeds = list(extra_seeds) + [rng.randrange(SIZE) for _ in range(trials)]
    for u in all_seeds:
        chars = generate_chars(u, num_chars)
        res = cvp_search(chars, prep)
        for x in res:
            if generate_chars(x, num_chars) != chars:
                false_pos += 1
                failures.append(f"false positive: anchor {x} for seed {u}")
        if len(res) > 1:
            multi += 1
        if u in res:
            found += 1
        else:
            not_found += 1
            tag = "edge" if u in extra_set else "random"
            failures.append(f"not found ({tag}): seed {u}, chars={chars}, got={res}")
    return {"found": found, "not_found": not_found, "false_pos": false_pos,
            "multi": multi, "failures": failures}


def generate_chars(u, length):
    """Characters produced starting from anchor seed u (u itself yields char0)."""
    out = [rand_int(u)]
    s = u
    for _ in range(1, length):
        s = next2(s)
        out.append(rand_int(s))
    return out

File: docs/js/charrss.js
// charrss.js -- client-side Reverse Seed Search for Melee's random character
// selection. Direct truncated-LCG / Hidden-Number-Problem seed reconstruction via
// a Schnorr-Euchner closest-vector (CVP) enumeration. Runtime port of charrss.py's
// cvp_search; the offline LLL reduction lives in tools/gen_charrss_constants.py and
// is baked into charrss_constants.js.
//
// Reconstructs the seed from 9 characters with no linear-combination budget
// constraint: the HNP lattice has no sum|c|<bound requirement, so 9 chars provides
// sufficient information (validated 5000/5000, no false positives --
// see RSS_IMPLEMENTATION.md).
//
// Implements searchForNewSeed() client-side. Given the first N characters
// (ints 0..24), returns the current RNG seed -- next^(2N-1)(anchor), identical
// to what rng.cpp locateCharSequence_ returned and to the successive-search
// endSeed -- so processSeed() is unchanged.
//
// NUMERICAL STRATEGY: correctness comes from regenerating the character sequence and
// comparing exactly at every enumeration leaf (rngInt semantics via util.js / next2,
// done in Number/Math.imul). The lattice/enumeration math is plain float64 -- it only
// PROPOSES candidate seeds, so double precision is sufficient (matches the Python
// reference exactly). No BigInt is needed.

import { rngAdv, rngInt } from './util.js';
import { CHARRSS_TABLES } from './charrss_constants.js';

// ---- LCG constants -------------------------------------------------------
const A = 214013;
const C = 2531011;
const BOUND = 25;
const HSIZE = 65536;
const MAXINT = 4294967295;       // 2^32 - 1
const SIZE = 4294967296;         // 2^32
// next2 = one character step = two LCG steps.  A*A and A*C+C are < 2^53, exact.
const A2 = (A * A) % SIZE;
const C2 = (A * C + C) % SIZE;

// Radius padding for the enumeration: the true error vector has squared norm <=
// sum of squared interval half-widths; we search a comfortably larger ball so float
// error can never prune the true seed (every leaf is exact-verified regardless).
const RADIUS_PAD = 1.5;

// next2(seed): two LCG steps, exact mod 2^32 via Math.imul.
function next2(seed) {
  return (Math.imul(seed, A2) + C2) >>> 0;
}

// ---- Interval helpers (bound=25) -- mirror charrss.py / rng.cpp CHAR_RANGES
function lowerBound(val) {
  return Math.floor((HSIZE / BOUND) * val + 1) * HSIZE;
}
function getLU(val) {            // [lo, hi] inclusive seed interval for char value
  if (val === 0) return [0, lowerBound(1) - 1];
  if (val === BOUND - 1) return [lowerBound(val), MAXINT];
  return [lowerBound(val), lowerBound(val + 1) - 1];
}

// b_k: affine offset of next2^k -- next2^k(s) = A2^k * s + b_k (mod 2^32).
// b_0 = 0, b_k = (b_{k-1}*A2 + C2) mod 2^32. Exact via Math.imul.
function charOffsets(n) {
  const b = new Array(n);
  b[0] = 0;
  for (let k = 1; k < n; k++) b[k] = (Math.imul(b[k - 1], A2) + C2) >>> 0;
  return b;
}

// regenerate the N characters produced from anchor u (u itself yields char0).
function generateChars(u, n) {
  const out = new Array(n);
  out[0] = rngInt(u, BOUND);
  let s = u;
  for (let k = 1; k < n; k++) { s = next2(s); out[k] = rngInt(s, BOUND); }
  return out;
}
function charsEqual(a, b) {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

// ---- Engine: CVP enumeration over a baked lattice for a fixed length ------
class CharRss {
  constructor(table) {
    this.n = table.n;
    this.gsStar = table.gsStar;     // b*_i  (n x n floats)
    this.gsDen = table.gsDen;       // ||b*_i||^2  (n floats)
    this.mu = table.mu;             // mu[j][i] = <R[j],b*_i>/||b*_i||^2
    this.r0 = table.r0;             // column 0 of R (anchor-seed coordinate)
    this.bvec = charOffsets(this.n);
    this.visited = 0;               // leaves verified by regeneration
  }

  // Returns { anchors: Number[] (ascending), visited: Number }.
  search(chars) {
    this.visited = 0;
    const n = this.n;
    if (chars.length !== n) {
      throw new Error(`charrss: expected ${n} characters, got ${chars.length}`);
    }
    const { gsStar, gsDen, mu, r0, bvec } = this;

    // Target (center of each char's seed interval, minus the affine offset) and
    // the squared-radius budget from the interval half-widths.
    const t = new Array(n);
    let halfw2 = 0;
    for (let k = 0; k < n; k++) {
      const [lo, up] = getLU(chars[k]);
      const center = Math.floor((lo + up) / 2);
      let tk = (center - bvec[k]) % SIZE;
      if (tk < 0) tk += SIZE;
      t[k] = tk;
      const hw = (up - lo) / 2;
      halfw2 += hw * hw;
    }
    const r2 = halfw2 * RADIUS_PAD;

    // Target coordinates in the Gram-Schmidt basis.
    const tstar = new Array(n);
    for (let i = 0; i < n; i++) {
      let acc = 0;
      for (let k = 0; k < n; k++) acc += t[k] * gsStar[i][k];
      tstar[i] = acc / gsDen[i];
    }

    const x = new Array(n).fill(0);
    const found = new Set();
    const self = this;

    // Schnorr-Euchner enumeration: pick coefficients from the most-significant GS
    // direction down; prune by accumulated squared distance; verify at the leaf.
    function rec(i, dacc) {
      if (i < 0) {
        let u = 0;
        for (let k = 0; k < n; k++) u += x[k] * r0[k];
        u >>>= 0;                                  // mod 2^32
        self.visited++;
        if (charsEqual(generateChars(u, n), chars)) found.add(u);
        return;
      }
      let s = 0;
      for (let j = i + 1; j < n; j++) s += x[j] * mu[j][i];
      const center = tstar[i] - s;
      const rem = r2 - dacc;
      if (rem < 0) return;
      const bound = Math.sqrt(rem / gsDen[i]);
      const loX = Math.floor(center - bound) - 1;
      const hiX = Math.ceil(center + bound) + 1;
      for (let xi = loX; xi <= hiX; xi++) {
        const d = xi - center;
        const dnew = dacc + gsDen[i] * d * d;
        if (dnew <= r2) { x[i] = xi; rec(i - 1, dnew); }
      }
    }
    rec(n - 1, 0);

    const anchors = Array.from(found).sort((a, b) => a - b);
    return { anchors, visited: this.visited };
  }
}

// ---- Public API ----------------------------------------------------------

// Build (and cache) an engine for a given character-sequence length.
const _engines = new Map();
function engineFor(numChars) {
  if (_engines.has(numChars)) return _engines.get(numChars);
  const table = CHARRSS_TABLES[numChars];
  if (!table) {
    throw new Error(`charrss: no baked table for ${numChars} characters ` +
      `(supported: ${Object.keys(CHARRSS_TABLES).join(', ')})`);
  }
  const eng = new CharRss(table);
  _engines.set(numChars, eng);
  return eng;
}

// searchForCharSeed(charSeq): the seed handed to processSeed -- next^(2N-1)(anchor),
// matching rng.cpp locateCharSequence_ and the successive-search endSeed convention.
// Returns -1 if no seed produces the sequence (a safe "not found": the verify step
// means a wrong candidate is never returned). Throws if the length has no baked table.
function searchForCharSeed(charSeq) {
  const N = charSeq.length;
  const eng = engineFor(N);
  const { anchors } = eng.search(charSeq);
  if (anchors.length === 0) return -1;
  if (anchors.length > 1) {
    // 9 chars uniquely pins the seed in practice; this is a safety net mirroring
    // the server (smallest match) + the tool's "enter more characters" model.
    console.warn(`charrss: ${anchors.length} candidate seeds; using smallest. ` +
      `Enter more characters to disambiguate.`);
  }
  let s = anchors[0];                 // anchor u: rngInt(u,25) == charSeq[0]
  const steps = 2 * N - 1;            // advance to the post-sequence seed
  for (let i = 0; i < steps; i++) s = rngAdv(s);
  return s;
}

// Internals exported for differential testing / asserting CHAR_RANGES parity.
export {
  searchForCharSeed,
  CharRss,
  engineFor,
  next2,
  lowerBound,
  getLU,
};
